### ISSUE TITLE
Nested hyper fields not saving

### ISSUE THREAD

**Issue by corneliusio** [verbb/hyper#217]
### Describe the bug

We've suddenly started having an issue where nested hyper fields are not updating. 

### Steps to reproduce

We currently have the following structure for a set of fields to allow for a site's main navigation:
* "Dropdown" (matrix type)
    * Label (plain text)
    * Menu (hyper field w/custom "dropdown" type for additional navigation layer)
      * Label (default Hyper link text input)
      * Links (hyper field)

To break it down fully, we have a matrix field with two types, "Dropdown" and "Menu Item". The "Dropdown" matrix type has a plain text field for the dropdown label in the menu and a Hyper field that can be populated with standard link types as well as a custom one that allows for an additional layer of navigation links. This custom link type uses the default "Link Text" input for it's label and has a different hyper field that populates this dropdown (this hyper field does *not* have an additional "dropdown" type so the nesting stops here). 

Now there are a couple things happening here that are odd:
1. The link value input of the custom Hyper "Dropdown" link type will automatically populate with a serialized array of ids that appear to correspond to the element ids of it's child hyper field. 
    1. I originally expected I'd need to add an event handler before save or validation to populate this input with something like "#" just to allow the field to pass validation, but this never ended up being the case due to this behavior.
3. No data for hyper related inputs or child hyper fields will update in the custom "Dropdown" link type *except* the type itself. This plays out as follows:
    1. If update the "Link Text" field for the "Dropdown" type, it reverts on save.
    2. If I add a child link to the "Dropdown" type, it's values are lost on save.
    3. If I update an existing URL child link to be an Entry type, the link will remain an Entry type after save but with no element attached.
    4. If I updated an existing child link Entry type to a URL, the URL value will be lost and a serialized array with a single id associated with the element assigned to the link before save will populate the link value input.

### Craft CMS version

5.5.3

### Plugin version

2.2.1

### Multi-site?

No

### Additional context

I'm more than happy to pass along project config files if it would help.

**Comment by engram-design** [comment]
Should be fixed for the next release. To get this early, run `composer require verbb/hyper:"dev-craft-5 as 2.2.1"`.

**Comment by corneliusio** [comment]
@engram-design Fantastic! That's fixed the issues for sure. I am a bit curious about this part:
> The link value input of the custom Hyper "Dropdown" link type will automatically populate with a serialized array of ids that appear to correspond to the element ids of it's child hyper field.
I originally expected I'd need to add an event handler before save or validation to populate this input with something like "#" just to allow the field to pass validation, but this never ended up being the case due to this behavior.

This is no longer occurring which makes sense to me, however when I go to implement the solution I'd mentioned it doesn't seem to work:
```php
Event::on(
    Entry::class,
    Entry::EVENT_BEFORE_VALIDATE,
    function (YiiModelEvent $event): void {
        /** @var Entry $block */
        $block = $event->sender;

        if ($block->type->handle === 'dropdown') {
            foreach ($block->menu as $menu_item) {
                if ($menu_item->label === 'Dropdown') {
                    // none of the following appear to work. no matter how 
                    // I try to programmatically populate this field, 
                    // I get a `Link cannot be blank.` validation error.
                    $menu_item->setAttributes([ 'linkValue' => '#' ], false);
                    $menu_item->setAttribute('linkValue', '#');
                    $menu_item->linkValue = '#';
                }
            }
        }
    },
);
```

Is there any way to bypass the validation for the link if needed? I know I'm stretching the functionality here a bit but I'd thought I'd gotten away with one since it was working for quite some time.

Also, while messing with this I came across another issue. The following line references an undefined variable `$item` which I'm assuming should be `$value`:
https://github.com/verbb/hyper/blob/craft-5/src/models/LinkCollection.php#L134

**Comment by engram-design** [comment]
While I don't have an example with Matrix, the basic premise is the same. It might be due to the fact you're not updating the field values properly. Here's an example for a top-level Hyper field.

```
Event::on(Entry::class, Entry::EVENT_BEFORE_VALIDATE, function($event) {
    $links = $event->sender->hyperField;
    $links[0]->linkValue = '#';

    $event->sender->setFieldValues([
        'hyperField' => $links,
    ]);
});
```

The value for a Hyper field is always going to be a `LinkCollection`, so you'll want to set the link values accordingly. Additionally, you'll want to use `setFieldValues` on the `$event->sender` variable (not a local variable like `$entry = $event->sender`).

One thing to bare in mind is that for a URL field, this won't pass validation, as that link type is strictly "it needs to be a URL, not URL-like". If you want to use anything, best to pick "Custom" as the link type.

Thanks for the typo heads-up! Fixed for the next release.

**Comment by corneliusio** [comment]
@engram-design Oh wow yeah, thanks for pointing that out. Had forgotten that the value is always a `LinkCollection` AND completely missed the obvious issue with assigning event sender to a local var. Sometime you just forget how to PHP. 😅

As for the URL field not allowing `#`, we tend to setup the built in "Custom" link type as the main URL type (just relabel it) for this exact reason. Everything's working great and as expected on our end now. Thanks again for the quick turnaround and gut check.

**Comment by engram-design** [comment]
Fixed in [2.2.2](https://github.com/verbb/hyper/releases/tag/2.2.2)
